kube-client/src/api/entry.rs

```diff
@@ -0,0 +1,394 @@
+//! API helpers for get-or-create and get-and-modify patterns
+//!
+//! [`Api::entry`] is the primary entry point for this API.
+
+// Import used in docs
+#[allow(unused_imports)] use std::collections::HashMap;
+use std::fmt::Debug;
+
+use kube_core::{params::PostParams, Resource};
+use serde::{de::DeserializeOwned, Serialize};
+
+use crate::{Api, Result};
+
+impl<K: Resource + Clone + DeserializeOwned + Debug> Api<K> {
+    /// Gets a given object's "slot" on the Kubernetes API, designed for "get-or-create" and "get-and-modify" patterns
+    ///
+    /// This is similar to [`HashMap::entry`], but [`Entry`] must be [`Entry::sync`]ed for changes to be persisted.
+    pub async fn entry<'a>(&'a self, name: &'a str) -> Result<Entry<'a, K>> {
+        Ok(match self.get_opt(name).await? {
+            Some(object) => Entry::Occupied(OccupiedEntry {
+                api: self,
+                object,
+                dirtiness: Dirtiness::Clean,
+            }),
+            None => Entry::Vacant(VacantEntry { api: self, name }),
+        })
+    }
+}
+
+#[derive(Debug)]
+/// A view into a single object, with enough context to create or update it
+///
+/// See [`Api::entry`].
+pub enum Entry<'a, K> {
+    /// An object that either exists on the server, or has been created locally (and is awaiting synchronization)
+    Occupied(OccupiedEntry<'a, K>),
+    /// An object that does not exist
+    Vacant(VacantEntry<'a, K>),
+}
+
+impl<'a, K> Entry<'a, K> {
+    /// Borrow the object, if it exists (on the API, or queued for creation using [`Entry::or_insert`])
+    pub fn get(&self) -> Option<&K> {
+        match self {
+            Entry::Occupied(entry) => Some(entry.get()),
+            Entry::Vacant(_) => None,
+        }
+    }
+
+    /// Borrow the object mutably, if it exists (on the API, or queued for creation using [`Entry::or_insert`])
```

Comment:
Since we are making `Entry` enum public, we should probably have language more like BTreeMap's `Entry` in these main methods on `Entry`. e.g. this one, maybe restructure to:

```suggestion
    /// Provides in-place mutable access to an occupied entry
    ///
    /// The entry is occupied if it exists on the API, or is queued for creation using [`Entry::or_insert`])
```

to have a more entry-language like header, and then we can explain what that means outside the header.

Reply:
Can we expect users to have read the enum-level docs already? I agree that the current wording is clunky, but I'd like to keep the emphasis on the unsaved/saved distinction (since this is the big difference compared to the `hash_map::Entry` API that the user is likely already familiar with).